Cleanup ocamlyacc

DemiMarie opened the pull request:
Hopefully less code means less maintenance burden.

DemiMarie committed:
ocamlyacc: refactor error messages

This deletes a bunch of duplicated code.

DemiMarie committed:
ocamlyacc: Fix -Wmissing-prototypes issues

This found no actual bugs, but it did identify significant amounts of
unused code.

DemiMarie committed:
ocamlyacc: Rip out register storage class specifier

The register storage class specifier is obsolete, as modern compilers
can do a much better job of register allocation.  The only exception is
the "register variables" extension supported by GCC, which OCaml
actually uses for its bytecode interpreter.

This change is specific to ocamlyacc because it could be done by a sed
call there.  Ripping out obsolete uses of "register" from the runtime
will require more work.

DemiMarie commented:
No change entry needed here

gasche commented:
Looks okay to me. The third commit has a mix of error-function refactoring and dead code removal, which makes it a bit less easy to review than it could be.

gasche commented on yacc/reader.c:
```diff
@@ -915,7 +764,7 @@ declare_tokens(int assoc)
             bp = get_name();
         else if (c == '\'' || c == '"')
-            bp = get_literal();
+            invalid_literal(lineno, line, cptr);
```
Actually I'm not sure I follow what is going on here, can you explain?
I don't understand the details of `get_literal`, but it seems to accept quoted string literals, and you are turning this into an error message here, why?

DemiMarie replied:
IIUC this feature comes from C, and generated code like

```c
enum Token {
   A = 'x',
};
```

This worked because in C, both enums and character constants are really just integers.  In OCaml this would generate something like

```ocaml
type token =
  | 'x'
```

which is an OCaml syntax error.

DemiMarie committed:
ocamlyacc: remove unused code

No change in behavior.

DemiMarie committed:
ocamlyacc: Forbid quoted literals in tokens and types

They would result in invalid OCaml code being generated, and removing
them allows for a bunch of code to be deleted.

DemiMarie committed:
ocamlyacc: fix indent

No change in behavior.

DemiMarie committed:
Merge branch 'trunk' into cleanup-ocamlyacc

gasche commented:
Merged, thanks!
